### ISSUE TITLE
carthage copy-frameworks produces "The file couldn’t be saved." error

### ISSUE THREAD

**Issue by adbqpa** [Carthage/Carthage#3056]
* carthage install method: [x] .pkg, [x] homebrew, [x] source
* `which carthage`:
* `carthage version`: 0.36.0
* `xcodebuild -version`: Build version 11E801a
* Are you using `--no-build`? No
* Are you using `--no-use-binaries`? No
* Are you using `--use-submodules`? No
* Are you using `--cache-builds`? Yes
* Are you using `--new-resolver`? No

**Actual outcome**
Xcode project build fails, because command "carthage copy-frameworks" returns non zero result, with error: "The file couldn’t be saved."
My investigation showed that the problem was a failed attempt to create a temporary directory due to reaching the maximum allowed limit of 1000 temporary directories. I found information about this limit on the forum https://developer.apple.com/forums/thread/128927
The complete error looks like this: Error Domain = NSCocoaErrorDomain Code = 512 "The file couldn’t be saved."
The code in which the error occurs:
```
public func createTemporaryDirectoryWithTemplate (_ template: String, destinationURL: URL) -> SignalProducer <URL, CarthageError> {
return SignalProducer {[base = self.base] () -> Result <String, CarthageError> in
      let temporaryDirectory: NSString
      do {
        temporaryDirectory = try base.url (
          for: .itemReplacementDirectory,
          in: .userDomainMask,
          appropriateFor: destinationURL,
          create: true
        )
        .path as NSString
      } catch {
>>>>>>> return .failure (.internalError (description: error.localizedDescription)) <<<<<<< error stopping the build here
      }
.............
```
**Expected outcome**
Carthage should clean up the created temporary folders at the end of the work in order not to reach this limit, or issue human-readable errors with guidance for further actions, instead of the system error "The file couldn’t be saved."

**Comment by Droppix** [comment]
same error
after some time, i have "The file couldn’t be saved."

```
/bin/sh -c /Users/xxxx/Library/Developer/Xcode/DerivedData/xxxxx-fmaccoypantpmwcbvalvgebzktwd/Build/Intermediates.noindex/xxxxx.build/Debug-iphonesimulator/xxxx.build/Script-2D6F48152518F1F300CCDBE0.sh

"The file couldn’t be saved.
Command PhaseScriptExecution failed with a nonzero exit code
```

The contents of Script-2D6F48152518F1F300CCDBE0.sh

```
#!/bin/sh
# Type a script or drag a script file from your workspace to insert its path.
/usr/local/bin/carthage copy-frameworks
```

Only one solution for me : Reboot MAC (clean derived data, don't work!)

**Comment by adbqpa** [comment]
> Only one solution for me : Reboot MAC (clean derived data, don't work!)

Clearing DerivedData should not help, the problem is in the $TMPDIR/TemporaryItems folder, hypothetically, it should be cleared when the computer is restarted, which in theory should help, but on my build machine, frequent reboots are not possible

**Comment by caobug** [comment]
Add to `run scripts`
```
rm -rf ${TMPDIR}/TemporaryItems/*carthage*
/usr/local/bin/carthage copy-frameworks
```

**Comment by PierreBrisorgueil** [comment]
@caobug tempo fix work for me :+1:

**Comment by kokluch** [comment]
+1

**Comment by ryecroft** [comment]
Thanks @caobug. I had to remove the first forward-slash from the command to get it to work:

```
rm -rf ${TMPDIR}TemporaryItems/*carthage*
/usr/local/bin/carthage copy-frameworks
```

Otherwise it was resulting in a path like `/var/folders/f1/5dssh6dx6rs_c9gbf039djww0000gn/T//TemporaryItems/`.

**Comment by AlexisQapa** [comment]
I ran into this issue and restarting solved the issue. The proposed fix in the comments didn't worked for me thought (with and without the /)

**Comment by maqoo** [comment]
I started to face this issue today morning and after I deleted all from `TemporaryItems` directory I was able to build successfully.

**Comment by kientux** [comment]
`rm -rf ${TMPDIR}TemporaryItems/*carthage*`

This command does not work on my machine, I tried to run it on Terminal and it gives `Operation not permitted` error.
Solution for who doesn't want to restart is to remove those temporary items manually by opening the `$TMPDIR` in Finder. Run this in Terminal:

`open $TMPDIR/TemporaryItems`

then just remove all directories created by Carthage (named `(A Document Being Saved By carthage)`)

**Comment by stepheng** [comment]
@kientux You can work around it by enabling Full Disk Access for Terminal (or, your build manager if you're using CI with remote machines).

**Comment by pjcau** [comment]
@kientux  I had the same issue, so you should enable Xcode App on  Full Disk Access.

**Comment by norbdev** [comment]
> macOS 10.15.6, Xcode 12.0.1

Only the (Xcode) "Full Disk Access" not working for me, have to add this:
```
rm -rf ${TMPDIR}TemporaryItems/*carthage*
/usr/local/bin/carthage copy-frameworks
```

**Comment by pjcau** [comment]
@norbdev I Use this lines
`echo "TMPDIR is ${TMPDIR}"
rm -rf ${TMPDIR}TemporaryItems/
/usr/local/bin/carthage copy-frameworks
`

**Comment by simonpierreroy** [comment]
At scale on CI machines, the limit is reached pretty fast. Most of my CI machines reached the 1000 limit inside `TemporaryItems` in a day! But indeed, removing item in "${TMPDIR}TemporaryItems/" will fix it for now. @adbqpa perfectly identified the issues https://github.com/Carthage/Carthage/issues/3056#issue-705971660, thank you.

**Comment by nnduc** [comment]
> Add to `run scripts`
> 
> ```
> rm -rf ${TMPDIR}/TemporaryItems/*carthage*
> /usr/local/bin/carthage copy-frameworks
> ```

Thank @caobug . It's work

**Comment by Droppix** [comment]
@norbdev 
I have same error, ```rm -rf ${TMPDIR}/TemporaryItems/*carthage*```work only on Terminal, but not in XCode.
Would you have found a solution

**Comment by raducanb** [comment]
@Droppix have you tried enabling full disk access for Xcode?
<img width="250" alt="Screenshot 2020-10-02 at 09 34 25" src="https://user-images.githubusercontent.com/4189954/94899048-e0e87880-0492-11eb-9c69-8f11646fafa7.png">

**Comment by Droppix** [comment]
@raducanb 
Sorry but yes, I knew it was checked! Thanks, it's ok, it works!

**Comment by SValchyshyn** [comment]
> Add to `run scripts`
> 
> ```
> rm -rf ${TMPDIR}/TemporaryItems/*carthage*
> /usr/local/bin/carthage copy-frameworks
> ```

If we do it this way on CI machine, it means fetching and building dependencies each time. Do we have a better approach? @simonpierreroy do you have such a problem?

**Comment by tmspzz** [comment]
@SValchyshyn you want to delete just `*carthage-lipo*`

**Comment by justAnotherDev** [comment]
i can confirm that the suggested solution fixes the error if presented during `carthage copy-frameworks`, however this error also sporadically presents itself during `carthage bootstrap`. 

if presented during `bootstrap` then the suggested solution does not help. the only working solution i have found is to reboot.

**Comment by geoherna** [comment]
Enabling full disk access for Xcode solved this issue for me.

**Comment by Johennes** [comment]
Is there any built-in fix for this coming up? If #3047 introduced this, could that change be reverted? As far as I understand #3047 tried to fix a sporadic issue. However, the new issue is deterministic now and impacts all our CI setups.

**Comment by aivcec** [comment]
This happens on my CI machine as well. None of the fixes above work. I have to manually restart the machine.

**Comment by Johennes** [comment]
@jdhealy / @tmspzz any thoughts on reverting #3047? Periodically logging into our CI servers to delete temporary files doesn't sound like a real solution.

**Comment by joelfischerr** [comment]
> This happens on my CI machine as well. None of the fixes above work. I have to manually restart the machine. 
> 
> Xcode 11.7, macOS 10.15.2

This looks like it could be caused by a CI runner without proper access rights. Have you checked whether it also has full disk access rights?

**Comment by trzeciak** [comment]
If you use TeamCity as the CI, its probably overrides $TMPDIR, therefore I try this workaround as a (Custom script) TeamCity Build Step:
```bash
echo 'import Foundation
let ls = Process()
ls.executableURL = URL(fileURLWithPath: "/usr/bin/env")
ls.arguments = ["bash", "-c", "set -x && rm -rfv " + NSTemporaryDirectory() + "TemporaryItems/*carthage*"]
try ls.run()
' | swift
```
_(I haven't found a simpler method to recover $TMPDIR)_

**Comment by anton-plebanovich** [comment]
This one works in Xcode run scripts:
```
for i in {1..1000}; do
    dir_name="${TMPDIR}TemporaryItems/(A Document Being Saved By carthage ${i})"
    [ -e "${dir_name}" ] && rm -rf "${dir_name}"
done
```

**Comment by aivcec** [comment]
Still happens to me and I don't believe its a permission issue.

I also see it happen on `carthage bootstrap`.

**Comment by ObjectiveCesar** [comment]
Any news on this one?

**Comment by ozgur** [comment]
> `rm -rf ${TMPDIR}TemporaryItems/*carthage*`
> 
> This command does not work on my machine, I tried to run it on Terminal and it gives `Operation not permitted` error.
> Solution for who doesn't want to restart is to remove those temporary items manually by opening the `$TMPDIR` in Finder. Run this in Terminal:
> 
> `open $TMPDIR/TemporaryItems`
> 
> then just remove all directories created by Carthage (named `(A Document Being Saved By carthage)`)

For those who don't want to do that manually, I've written a simple python script in order to navigate around that nasty `Operation not permitted` error:

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import glob
import os
import shutil
import tempfile

if __name__ == "__main__":
    tmpdir = os.path.join(tempfile.gettempdir(), "TemporaryItems")
    for f in glob.glob(f'{tmpdir}/*carthage*/'):
        shutil.rmtree(os.path.join(tmpdir, f), ignore_errors=True)
```

then on terminal
```bash
$ python /path/to/script.py
```
should clean the folders inside tmp folder created by Carthage.

PS: Don't forget to grant write permission: `sudo chmod -R +w "${TMPDIR}"`

**Comment by clarissaj** [comment]
FYI it looks like this issue is fixed in the latest release!! ([1.36.1](https://github.com/Carthage/Carthage/releases/tag/0.36.1)) 🎉🎉🎉🎉

**Comment by ObjectiveCesar** [comment]
> FYI it looks like this issue is fixed in the latest release!! ([0.36.1](https://github.com/Carthage/Carthage/releases/tag/0.36.1)) 🎉🎉🎉🎉

That would be awesome. I‘ll try to verify that on Monday. Thanks for the heads-up!
